lib/ffmpeg/encoding_options.rb

```diff
@@ -9,27 +9,90 @@ def initialize(options = {}, prefix_options = {})
 
     def to_s
       params = collect do |key, value|
-        send("convert_#{key}", value) if value && supports_option?(key)
+        attempt_self_call(key, value)
       end
 
       prefix_params = @prefix_options&.map do |key, value|
-        send("convert_#{key}", value) if value && supports_option?(key)
+        attempt_self_call(key, value)
       end
 
       # codecs should go before the presets so that the files will be matched successfully
       # all other parameters go after so that we can override whatever is in the preset
-      input   = params.select { |p| p =~ /^\-i / }
-      seek    = params.select {|p| p =~ /\-ss/ }
-      codecs  = params.select { |p| p =~ /codec/ }
-      presets = params.select { |p| p =~ /\-.pre/ }
-      other   = params - codecs - presets - input - seek
-      params  = prefix_params + seek + input + codecs + presets + other
+      inputs                    = params.select { |p| p =~ /\-i / }
+      seek                      = params.select {|p| p =~ /\-ss/ }
+      codecs                    = params.select { |p| p =~ /codec/ }
+      presets                   = params.select { |p| p =~ /\-.pre/ }
+      contains_complex_filter   = params.any? { |p| p =~ /\-filter_complex / }
+
+      other   = params - codecs - presets - inputs - seek
+      params  = prefix_params + seek + inputs + codecs + presets + other
+
+      num_inputs = inputs.first&.scan(/(?=\-i)/)&.count || 0
+      if num_inputs > 1 && !contains_complex_filter
+        multi_input_output_filter = "-filter_complex \"#{default_multi_input_complex_filter(num_inputs)}\" -map \"[v]\" -map \"[a]\""
+        params.push(multi_input_output_filter)
+      end
 
       params_string = params.join(" ")
       params_string << " #{convert_aspect(calculate_aspect)}" if calculate_aspect?
+
       params_string
     end
 
+    def to_s_minimal
```

Comment:
Why do we have two `to_s` methods? There's a lot of duplication between the two

Reply:
These are serving different purposes and not easily/cleanly abstracted into helper methods

Reply:
Ok. Can you please add code comments that explain the differences?

Reply:
Sure, added.